tests/test_rtm.py

```diff
@@ -0,0 +1,345 @@
+                tracker.handle_all_messages(event, message)
+
+            @rtm.handle_message('^ham')
+            @rtm.handle_message('^spam ham')
+            def handle_spam_messages(self, event, message):
+                tracker.handle_spam_messages(event, message)
+
+        service_runner(Service, events)
+
+        assert (
+            [call(event) for event in events] ==
+            tracker.handle_all_events.call_args_list)
+
+        assert (
+            [
+                call(event) for event in events
+                if event.get('type') == 'presence_change'
+            ] ==
+            tracker.handle_presence_changes.call_args_list)
+
+        assert (
+            [
+                call(event, event.get('text')) for event in events
+                if event.get('type') == 'message'
+            ] ==
+            tracker.handle_all_messages.call_args_list)
+
+        assert (
+            [
+                call(make_message_event(text='spam ham'), 'spam ham'),
+                call(make_message_event(text='ham spam'), 'ham spam'),
+            ] ==
+            tracker.handle_spam_messages.call_args_list)
+
+
+@patch('nameko_slack.rtm.SlackClient')
+def test_handlers_do_not_block(
+    SlackClient, container_factory, config, tracker
+):
+
+    work_1 = Event()
+    work_2 = Event()
+
+    class Service:
+
+        name = 'sample'
+
+        @rtm.handle_event
+        def handle_1(self, event):
+            work_1.wait()
+            tracker.handle_1(event)
+
+        @rtm.handle_event
+        def handle_2(self, event):
+            work_2.wait()
+            tracker.handle_2(event)
+
+    events = [{'spam': 'ham'}]
+
+    def rtm_read():
+        if events:
+            return [events.pop(0)]
+        else:
+            return []
+
+    SlackClient.return_value.rtm_read.side_effect = rtm_read
+    container = container_factory(Service, config)
+    container.start()
+
+    try:
+
+        # both handlers are still working
+        assert (
+            [] ==
+            tracker.handle_1.call_args_list)
+        assert (
+            [] ==
+            tracker.handle_2.call_args_list)
+
+        # finish work of the second handler
+        work_2.send()
+        sleep(0.1)
+
+        # second handler is done
+        assert (
+            [] ==
+            tracker.handle_1.call_args_list)
+        assert (
+            [call({'spam': 'ham'})] ==
+            tracker.handle_2.call_args_list)
+
+        # finish work of the first handler
+        work_1.send()
+        sleep(0.1)
+
+        # first handler is done
+        assert (
+            [call({'spam': 'ham'})] ==
+            tracker.handle_1.call_args_list)
+        assert (
+            [call({'spam': 'ham'})] ==
+            tracker.handle_2.call_args_list)
+    finally:
+        if not work_1.ready():
+            work_1.send()
+        if not work_2.ready():
+            work_2.send()
+
+    container.stop()
```

Comment:
The `container_factory` fixture automatically stops any started containers on teardown.